### PR TITLE
Add support to write to ABFS filesystems.

### PR DESCRIPTION
**Pull request by gaoyangxiaozhu**
This change adds support to write from Abfs (Azure Gen2 Storage).

It's part of work related to issue https://github.com/facebookincubator/velox/issues/6415

### CODE REVIEW

File: velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp [hunk 1]
```diff
@@ -213,6 +216,155 @@ uint64_t AbfsReadFile::getNaturalReadSize() const {
   return impl_->getNaturalReadSize();
 }
 
+class BlobStorageFileClient final : public IBlobStorageFileClient {
+ public:
+  BlobStorageFileClient(const DataLakeFileClient client)
+      : client_(std::make_unique<DataLakeFileClient>(client)) {}
+
+  void Create() override {
```
Comment (kgpai) [inline velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp:224]: These function names dont  follow Velox convention : https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#naming-conventions

File: scripts/setup-adapters.sh [hunk 2]
```diff
@@ -62,7 +62,9 @@ function install_gcs-sdk-cpp {
 }
 
 function install_azure-storage-sdk-cpp {
-  github_checkout azure/azure-sdk-for-cpp azure-storage-blobs_12.8.0
+  vcpkg_commit_id=7a6f366cefd27210f6a8309aed10c31104436509
```
Comment (kgpai) [inline scripts/setup-adapters.sh:65]: Curious , would there be problems with previous installs of the azure storage library using this same script (i.e conflicts or picking up the wrong version etc)..

File: CMakeLists.txt [hunk 3]
```diff
@@ -206,7 +206,7 @@ if(VELOX_ENABLE_ABFS)
   if(AZURESDK_ROOT_DIR)
     list(APPEND CMAKE_PREFIX_PATH ${AZURESDK_ROOT_DIR})
   endif()
-  find_package(azure-storage-blobs-cpp CONFIG REQUIRED)
+  find_package(azure-storage-files-datalake-cpp CONFIG REQUIRED)
```
Comment (kgpai) [inline CMakeLists.txt:209]: nit: should it be versioned ?

**Comment by kgpai** [comment]
cc: @majetideepak

File: CMakeLists.txt (hunk 3 above)
Comment (gaoyangxiaozhu) [reply CMakeLists.txt:209]: currently i don't see versioned is needed here when tried 12.6+ due to no broken chagne for used APIs

**Comment by majetideepak** [comment]
@gaoyangxiaozhu Can you move the dependency change to a separate PR?

File: velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.h [hunk 4]
```diff
@@ -36,8 +36,8 @@ static const std::string AzuriteAccountKey{
     "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw=="};
 static const std::string AzuriteABFSEndpoint = fmt::format(
     "abfs://{}@{}.dfs.core.windows.net/",
-    AzuriteAccountName,
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.h:39]: Why did this not break anything before?

File: CMakeLists.txt (hunk 3 above)
Comment (majetideepak) [reply CMakeLists.txt:209]: Don't we need the `azure-storage-blobs-cpp` package?

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ * However, the Azurite Simulator does not yet support the DFS endpoint.
+ * (For more information, see https://github.com/Azure/Azurite/issues/553 and
+ * https://github.com/Azure/Azurite/issues/409).
+ * You can find a comparison between DFS and Blob endpoints here:
+ * https://github.com/Azure/Azurite/wiki/ADLS-Gen2-Implementation-Guidance
+ *
+ * To facilitate unit testing of file write scenarios, we define the
+ * IBlobStorageFileClient here, which can be mocked during testing.
+ */
+class IBlobStorageFileClient {
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h:49]: The definition can go to the test file. We can forward declare here.

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp [hunk 6]
```diff
@@ -15,29 +15,86 @@
...
 constexpr int kOneMB = 1 << 20;
 static const std::string filePath = "test_file.txt";
 static const std::string fullFilePath =
     facebook::velox::filesystems::test::AzuriteABFSEndpoint + filePath;
 
+// A mocked blob storage file client backend with local file store.
+class MockBlobStorageFileClient : public IBlobStorageFileClient {
+ public:
+  MockBlobStorageFileClient() {
+    char tempFileName[] = "/tmp/velox_abfs_test_XXXXXX";
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:49]: Use `createTempDirectory` API.

File: CMakeLists.txt (hunk 3 above)
Comment (majetideepak) [reply CMakeLists.txt:209]: I am guessing `azure-storage-files-datalake-cpp` package includes both  libraries.

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu Can you move the dependency change to a separate PR?

sure, will do

**Comment by gaoyangxiaozhu** [comment]
> > @gaoyangxiaozhu Can you move the dependency change to a separate PR?
> 
> sure, will do

hey @majetideepak i send seperate PR for dependency change as you suggested https://github.com/facebookincubator/velox/pull/7688 could you help reivew ?

File: velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.h (hunk 4 above)
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.h:39]: @majetideepak , because AzuriteAccountName and AzuriteContainerName both are `test`

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp (hunk 6 above)
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:49]: hey @majetideepak  this is used to create a temp file not a temp directory

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h (hunk 5 above)
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h:49]: @majetideepak  thank you! So you mean we can put it in such as `abfsWritefileTest.h` and then let `abfsWriteFile.h` foward the declare to use it ? But i am wondering does it a good way to let business code depends on defination from test file ? Do we already have any implemented code example referenece for same situation (using defination defined in test file from business code file ?)

File: velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp [hunk 7]
```diff
@@ -213,6 +216,155 @@ uint64_t AbfsReadFile::getNaturalReadSize() const {
...
+  const std::string connectStr_;
+  std::string fileSystem_;
+  std::string fileName_;
+  std::shared_ptr<IBlobStorageFileClient> blobStorageFileClient_;
+
+  uint64_t position_ = -1;
+  std::atomic<bool> closed_{false};
+};
+
+AbfsWriteFile::AbfsWriteFile(
```
Comment (zhli1142015) [inline velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp:337]: Is there any specific reason we implement several classes in the same cpp file, can we move this to a separate cpp file?

Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp:337]: fore read part please check this comment for why https://github.com/facebookincubator/velox/pull/6675#discussion_r1358142488

for write part `abfsWriteFile` of this PR , I just follow how other connector do, but I think write part can be move to seperate cpp if needed.

Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp:337]: We could add separate CPP files for each class. The intention of the comment is to only reduce the number of headers.

**Comment by majetideepak** [comment]
@gaoyangxiaozhu is this ready for review?

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu is this ready for review?

yes @majetideepak please, thank you!

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu is this ready for review?
just ping @majetideepak to help review if have time

**Comment by majetideepak** [comment]
@gaoyangxiaozhu I will get to this tomorrow. Sorry for the delay.

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu I will get to this tomorrow. Sorry for the delay.

it's ok @majetideepak thank you very much!

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h (hunk 5 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h:49]: I see your point. `IBlobStorageFileClient` interface is implemented inside both `AbfsWriteFile.cpp` and `AbfsFileSystemTest.cpp`. We have to keep it here then.

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  void flush() override;
+
+  /// Write the data by append mode.
+  void append(std::string_view data) override;
+
+  /// Close the file.
+  void close() override;
+
+  /// mainly for test purpose.
+  void setFileClient(std::shared_ptr<IBlobStorageFileClient> fileClient);
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h:84]: `testingSetFileClient`

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+  /// Flush the data.
+  void flush() override;
+
+  /// Write the data by append mode.
+  void append(std::string_view data) override;
+
+  /// Close the file.
+  void close() override;
+
+  /// mainly for test purpose.
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h:83]: /// Used by tests to override the FileSystem client.

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h"
+#include <azure/storage/files/datalake.hpp>
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:18]: nit: empty line above.

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+ private:
+  const std::string path_;
+  const std::string connectStr_;
+  std::string fileSystem_;
+  std::string fileName_;
+  std::shared_ptr<IBlobStorageFileClient> blobStorageFileClient_;
+
+  uint64_t position_ = -1;
+  std::atomic<bool> closed_{false};
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:136]: I don't think atomic is needed here since the assumption is that `AbfsWriteFile` is not thread-safe.

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+  void append(std::string_view data) {
+    VELOX_CHECK(!closed_, "File is not open");
+    if (data.size() == 0) {
+      return;
+    }
+    append(data.data(), data.size());
+  }
+
+  uint64_t size() const {
+    auto properties = blobStorageFileClient_->getProperties();
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:117]: `return blobStorageFileClient_->getProperties().FileSize;`

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp (hunk 6 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:49]: There is a `TempFilePath` API as well.
`/velox/exec/tests/utils/TempFilePath.h`

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+uint64_t AbfsWriteFile::size() const {
+  return impl_->size();
+}
+
+void AbfsWriteFile::setFileClient(
+    std::shared_ptr<IBlobStorageFileClient> fileClient) {
+  impl_->setFileClient(std::move(fileClient));
+}
+} // namespace facebook::velox::filesystems::abfs
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:169]: nit: empty line below.

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+    }
+    append(data.data(), data.size());
+  }
+
+  uint64_t size() const {
+    auto properties = blobStorageFileClient_->getProperties();
+    return properties.FileSize;
+  }
+
+  void append(const char* buffer, size_t size) {
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:121]: simplify as
```
blobStorageFileClient_->append(
        reinterpret_cast<const uint8_t*>(buffer), size, position_);
position_ += size;
```

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+    blobStorageFileClient_->create();
+  }
+
+  /// mainly for test purpose.
+  void setFileClient(
+      std::shared_ptr<IBlobStorageFileClient> blobStorageManager) {
+    blobStorageFileClient_ = std::move(blobStorageManager);
+  }
+
+  bool exist() {
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:81]: This is only used once. Can we inline the body at the call site?

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp [hunk 16]
```diff
@@ -15,29 +15,86 @@
...
+using namespace Azure::Storage::Files::DataLake;
 using ::facebook::velox::common::Region;
 
 constexpr int kOneMB = 1 << 20;
 static const std::string filePath = "test_file.txt";
 static const std::string fullFilePath =
     facebook::velox::filesystems::test::AzuriteABFSEndpoint + filePath;
 
+// A mocked blob storage file client backend with local file store.
+class MockBlobStorageFileClient : public IBlobStorageFileClient {
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:46]: Consider moving this to another file to simplify.

**Review by majetideepak — COMMENTED** [review]
@gaoyangxiaozhu some comments!

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp (hunk 16 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:46]: Consider adding `AbfsFileSystemWriteTest.cpp` and renaming the existing test file to `AbfsFileSystemReadTest.cpp` if feasible.

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp (hunk 6 above)
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp]: updated. thanks @majetideepak

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp (hunk 15 above)
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:79]: hey @majetideepak   firstly, i super thank you for helping review. And recent weeks I am busying with internel project so didn't have change to resolve your comments. Hope you can understand,

And for this comment, you mean I can just move `blobStorageFileClient_ = std::move(blobStorageManager);` this part directly of part of body of `AbfsWriteFile::testingSetFileClient`  ?

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp (hunk 16 above)
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:46]: > Consider adding `AbfsFileSystemWriteTest.cpp` and renaming the existing test file to `AbfsFileSystemReadTest.cpp` if feasible.

hey @majetideepak the test case not only have `read` and `write` part, also inlcuding other ops as `cred` check, `list/rename/mkdir` etc , so just split to `ReadTest` and `WriteTest` looks not a good refactor.

I would like to keep still one test file in this PR, and send antoher seperate PR do refactor if needed.

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp (hunk 10 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:17]: We need to add a blank line between the various inlude categories. See 
https://github.com/facebookincubator/velox/discussions/7723#discussioncomment-7687271

Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:17]: oh, got it! thank you @majetideepak . Resolved

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp (hunk 15 above)
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp]: done!

**Comment by gaoyangxiaozhu** [comment]
@majetideepak  the build is down but if you have time please help check if any other part of this PR need refactor. thank you!

**Comment by majetideepak** [comment]
@gaoyangxiaozhu I will take a look. We also noticed a build problem with the Azure library reported here https://github.com/facebookincubator/velox/issues/8622 Can you take a look?

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp (hunk 16 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:46]: Can you add the `MockBlobStorageFileClient` introduced here to a separate .h/.cpp

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu I will take a look. We also noticed a build problem with the Azure library reported here #8622 Can you take a look?

sure @majetideepak will check

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+        return false;
+      } else {
+        throwStorageExceptionWithOperationDetails("GetProperties", path_, e);
+      }
+    }
+  }
+
+  const std::string path_;
+  const std::string connectStr_;
+  std::string fileSystem_;
```
[the remark this answers is not in the document]
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:129]: we currerntly use those for auth token provider part in our internal velox repository but that part still not been send to OSS yet,  let me remove those properties now

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+  std::string fileName_;
```
[the remark this answers is not in the document]
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:130]: removed all useless properties

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+      if (e.StatusCode == Azure::Core::Http::HttpStatusCode::NotFound) {
+        return false;
+      } else {
+        throwStorageExceptionWithOperationDetails("GetProperties", path_, e);
+      }
+    }
+  }
+
+  const std::string path_;
+  const std::string connectStr_;
```
[the remark this answers is not in the document]
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:128]: sure

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <folly/executors/IOThreadPoolExecutor.h>
+#include <iostream>
+#include "AbfsUtil.h"
+#include "folly/io/Cursor.h"
```
[the remark this answers is not in the document]
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h:21]: oh, the parallel write part code not included in this PR, so let me remove this headers.

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <folly/executors/IOThreadPoolExecutor.h>
+#include <iostream>
+#include "AbfsUtil.h"
```
[the remark this answers is not in the document]
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h:20]: removed

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <folly/executors/IOThreadPoolExecutor.h>
```
[the remark this answers is not in the document]
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h:18]: not need anymore @majetideepak , just removed

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp (hunk 16 above)
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp]: sure @majetideepak missed that part before

File: velox/connectors/hive/storage_adapters/abfs/tests/MockBlobStorageFileClient.h [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h"
+#include "velox/exec/tests/utils/TempFilePath.h"
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/tests/MockBlobStorageFileClient.h:18]: empty line above

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp [hunk 24]
```diff
@@ -185,13 +228,37 @@ TEST_F(AbfsFileSystemTest, missingFile) {
...
 }
 
-TEST_F(AbfsFileSystemTest, openFileForWriteNotImplemented) {
-  auto hiveConfig = AbfsFileSystemTest::hiveConfig(
-      {{"fs.azure.account.key.test.dfs.core.windows.net",
-        azuriteServer->connectionStr()}});
-  auto abfs = std::make_shared<filesystems::abfs::AbfsFileSystem>(hiveConfig);
+TEST_F(AbfsFileSystemTest, OpenFileForWriteTest) {
+  const std::string abfsFile =
+      facebook::velox::filesystems::test::AzuriteABFSEndpoint + "writetest.txt";
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:233]: `filesystems::test::AzuriteABFSEndpoint` should be sufficient.

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp [hunk 25]
```diff
@@ -185,13 +228,37 @@ TEST_F(AbfsFileSystemTest, missingFile) {
...
+  auto mockClient = std::make_shared<
+      facebook::velox::filesystems::test::MockBlobStorageFileClient>(
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:235]: `filesystems::test::MockBlobStorageFileClient`

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp [hunk 26]
```diff
@@ -185,13 +228,37 @@ TEST_F(AbfsFileSystemTest, missingFile) {
...
+      facebook::velox::filesystems::test::MockBlobStorageFileClient());
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:236]: `filesystems::test::MockBlobStorageFileClient()`

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+ */
+
+#include "velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h"
+
+#include <azure/storage/files/datalake.hpp>
+
+namespace facebook::velox::filesystems::abfs {
+class BlobStorageFileClient final : public IBlobStorageFileClient {
+ public:
+  BlobStorageFileClient(const DataLakeFileClient client)
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:24]: `BlobStorageFileClient(const std::unique_ptr<DataLakeFileClient> client)`

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+    if (position_ != -1) {
+      return;
+    }
+    position_ = 0;
+  }
+
+  void initialize() {
+    if (!blobStorageFileClient_) {
+      auto abfsAccount = AbfsAccount(path_);
+      auto fileClient = DataLakeFileClient::CreateFromConnectionString(
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:66]: auto fileClient = std::make_unique<DataLakeFileClient>(DataLakeFileClient::CreateFromConnectionString(connectStr_, abfsAccount.fileSystem(), abfsAccount.filePath()));

**Review by majetideepak — COMMENTED** [review]
@gaoyangxiaozhu some more comments. Thanks!

File: velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp [hunk 29]
```diff
@@ -250,4 +248,13 @@ std::unique_ptr<ReadFile> AbfsFileSystem::openFileForRead(
   abfsfile->initialize();
   return abfsfile;
 }
+
+std::unique_ptr<WriteFile> AbfsFileSystem::openFileForWrite(
+    std::string_view path,
+    const FileOptions& /*unused*/) {
+  auto abfsfile = std::make_unique<AbfsWriteFile>(
+      std::string(path), impl_->connectionString(std::string(path)));
+  abfsfile->initialize();
```
[the remark this answers is not in the document]
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp:257]: yes, @majetideepak , mainly for we can have chance to do file initialization related ops after fileClient been updated if needed.  And right now only used in test scenario https://github.com/gayangya/velox/blob/5a21b990d9b9d2cae7b8b9cd7279074520d682e6/velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp#L80C1-L81C28

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp [hunk 30]
```diff
@@ -72,13 +75,53 @@ class AbfsFileSystemTest : public testing::Test {
...
+          client) {
+    auto abfsfile =
+        std::make_unique<facebook::velox::filesystems::abfs::AbfsWriteFile>(
+            std::string(path), azuriteServer->connectionStr());
+    abfsfile->testingSetFileClient(client);
+    abfsfile->initialize();
+    return abfsfile;
+  }
+
+  static std::string generateRandomData(int size) {
```
[the remark this answers is not in the document]
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:85]: @majetideepak this method giving a flexibility to generate any size data. I think generating ramdom content is fine, since we can keep a duplicate one in mem and veiry with data got from actually read.

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp (hunk 27 above)
Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp]: hey @majetideepak using const unique_ptr will causing can't assign to `_client` via `_client(client)`, i update to use shared_ptr, hope that also acceptable.

Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp]: @gaoyangxiaozhu You have use `std::move()` for assigning a unique pointer. Let's stick with `unique_ptr`. Thanks.

Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp]: hello @majetideepak but when i change to below

```
  BlobStorageFileClient(const std::unique_ptr<DataLakeFileClient> client)
      : client_(std::move(client)) {}
```

it shows error

```
function "std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp> &) [with _Tp=Azure::Storage::Files::DataLake::DataLakeFileClient, _Dp=std::default_delete<Azure::Storage::Files::DataLake::DataLakeFileClient>]" (declared at line 414 of "/usr/include/c++/9/bits/unique_ptr.h") cannot be referenced -- it is a deleted function
```

Comment (gaoyangxiaozhu) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp]: I eventually remove `const` it pass the build. 
But I don't know what's the reason here

File: velox/common/memory/tests/MemoryAllocatorTest.cpp [hunk 31]
```diff
@@ -123,25 +123,7 @@ class MemoryAllocatorTest : public testing::TestWithParam<int> {
   /// Returns the virtual and resident sizes of the process in 4K pages. Only
   /// defined for Linux.
   std::optional<ProcessSize> processSize() {
```
Comment (majetideepak) [inline velox/common/memory/tests/MemoryAllocatorTest.cpp:125]: Can we revert these changes?

**Review by majetideepak — COMMENTED** [review]
@gaoyangxiaozhu, Just one comment.

File: velox/common/memory/tests/MemoryAllocatorTest.cpp (hunk 31 above)
Comment (gaoyangxiaozhu) [reply velox/common/memory/tests/MemoryAllocatorTest.cpp]: oh god! @majetideepak It for unblock local build but I shouldn't be committed. Sorry for this mistake. Looks I wasn't focused enough on coding at home during the Spring Festival holiday.

Reverted

**Review by majetideepak — APPROVED** [review]
Thanks, @gaoyangxiaozhu

**Comment by majetideepak** [comment]
@pedroerp can you help merge this? Thanks.

**Review by kgpai — APPROVED** [review]
